Tweaks for UIA keyboard shortcut key retriever routine. re #6779, #6790

josephsl opened the pull request:
Resolves the following issues:

* #6779: string concatenation error when focus event is fired or when attempting to access certain UIA controls via object navigation by checking if access key is None or not.
* #6790: Two spaces are no longer inserted if only access key or accelerator key is returned by UIA.

Suggested What's New entries:
Bug fixes:

* NVDA will no longer fail to navigate to or announce UIA controls where no keyboard shortcuts are defined. (#6779)
* In certain UIA controls, two empty spaces are no longer added in keyboard shortcut information if the control does not have more than one keyboard shortcut. (#6790)

Thanks.

josephsl committed:
UIA keyboard shortcut solutions, including concatenation error and no longer adding two spaces blindly if access key is there but accelerator isnt. re #6779, #6790.

For 6779: there's no check for the case where access key may return NULL (None), thus take care of this by forcefully setting access key to an empty string.
For 6790: don't add two spaces blindly if access key is there but accelerator isn't - keyboard shortcut is queried not only in Visual Studio, but also in other programs such as Outlook and many others.
Overall code: made it more Pythonic (for instance, using str.join instead of string concatenation).

josephsl committed:
UIA keyboard shortcuts: insert two apces if and only if both access key and accelerator are present, return longest string otherwise. re #6790.

A Python trick: when checking for logical and for two strings, it'll return a string (an empty string if one or both are truly empty). If logical and says 'false', then just return the longest string as the keyboard shortcut, otherwise insert two spaces.

jcsteh commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -767,19 +767,19 @@ def _get_description(self):
-		ret = ""
 		try:
-			ret += self.UIAElement.currentAccessKey
+			accessKey = self.UIAElement.currentAccessKey
```
It probably shouldn't happen, but it's possible that this could itself return None (rather than throwing an exception). Since we're being safe, we should probably handle that.

jcsteh commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -767,19 +767,19 @@ def _get_description(self):
+		# #6790: Do not add two spaces unless both access key and accelerator are present in order to not waste string real estate.
+		# If not, just return the longest string.
+		keyboardShortcuts = (accessKey, acceleratorKey)
+		return "  ".join(keyboardShortcuts) if accessKey and acceleratorKey else max(keyboardShortcuts)
```
This is clever, but I think a bit tricky to read. Since we're creating a tuple anyway, I suggest just creating a list at the start and appending the values if they aren't None/exception.

josephsl committed:
UIA obj.keyboardShortcut: build keyboard shortucts list early for readability, catch rare cases where access key itself could be NULL. re #6779, #6790.

Reviewed by Jamie Teh 9NV Access): build keyboard shortcuts list early for readability. Also, take care of rare situations where access key itself could return NULL (None).

jcsteh reviewed (changes requested):
If fetching either key throws an exception, the variable won't be set at all, so you'll get an AttributeError. Also, the max() thing is still hard to read IMO. I'd just do something like this for both keys:

```
shortcuts = []
try:
	accessKey = ...
	if accessKey:
		shortcuts.append(accessKey)
except COMError:
	pass
# Same for accelerator...
...
return "  ".join(shortcuts)
```

This way, the list only contains either one or two non-empty strings and spaces only separate if there are two. It's also reasonably readable.

josephsl committed:
UIA objects.keyboardShortcuts: readability improvements, catch attribute error early, append to shortcuts list early. re #6779, #6790.

Reviewed by Jamie Teh (NV Access): we still want to catch NULL, so catch AttributeError exception early. Also, append access key and accelerator as soon as possible 9if they are even defined). Lastly, simplified the return statement to check for presence of access key and/or accelerator in the shortcuts list (len is at least 1), and return an an empty string if no info found. These changes shuld hopefully improve readability.
